### PR TITLE
Fix servo bq

### PR DESCRIPTION
**Pull request by IJDykeman**

### CODE REVIEW

File: embedded/servo_driver/servo_driver.cc [hunk 1]
```diff
@@ -14,17 +14,26 @@ namespace {
...
 constexpr int MAX_COUNTS = 2935;
+constexpr int PWM_FREQUENCY = 330;
 }  // namespace
 
-ServoDriver::ServoDriver(const int channel,
-                         const std::shared_ptr<PwmDriver> &pwm_driver,
-                         const std::string &config_path)
-    : channel_(channel), pwm_driver_(pwm_driver), config_path_(config_path) {
+ServoDriver::ServoDriver(const std::string &config_path) : config_path_(config_path) {
+  int i2cHandle = i2c_open("/dev/i2c-1");
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.cc]: add a TODO, allow this to be configurable

File: infrastructure/scripts/jet_run.sh [hunk 2]
```diff
@@ -50,10 +50,11 @@ WEBCAM_DEFAULT_DEV_PATH=/dev/video0
 if [ -d "$WEBCAM_DEFAULT_DEV_PATH" ]; then
   ATTACH_WEBCAM_DEVICE=--device=/dev/video0:/dev/video0
 fi
+SERVO_DEVICE=--device=/dev/i2c-1
 
 xhost +
 
-CONTAINER_ID=$(docker run -it -d -v $JET_REPO_PATH:/jet -v $JET_REPO_PATH/logs:/logs -v /tmp/.X11-unix:/tmp/.X11-unix -e LOG_BASE_PATH=/logs/ -e DISPLAY -e NO_AT_BRIDGE=1 -e MQTT_ADDRESS=$MQTT_ADDRESS $ATTACH_WEBCAM_DEVICE --net=host --privileged $FULL_IMAGE_NAME bash)
+CONTAINER_ID=$(docker run -it -d -v $JET_REPO_PATH:/jet -v $JET_REPO_PATH/logs:/logs -v /tmp/.X11-unix:/tmp/.X11-unix -e LOG_BASE_PATH=/logs/ -e DISPLAY -e NO_AT_BRIDGE=1 -e MQTT_ADDRESS=$MQTT_ADDRESS $ATTACH_WEBCAM_DEVICE $SERVO_DEVICE --net=host --privileged $FULL_IMAGE_NAME bash)
```
Comment (jpanikulam) [inline infrastructure/scripts/jet_run.sh]: What does `$SERVO_DEVICE` do?

File: embedded/servo_bq/set_servo_message.hh [hunk 3]
```diff
@@ -1,13 +1,23 @@
...
+// struct ServoCommand{
+// 	int servo_index;
+// 	float target_percentage;
+// 	SERIALIZABLE_STRUCTURE(ServoCommand, servo_index, target_percentage);
+// };
+
 struct SetServoMessage : Message {
-  float target_angle; //TODO units
+  std::vector<int> servo_indices;
+  std::vector<float> target_percentages;
```
Comment (jpanikulam) [inline embedded/servo_bq/set_servo_message.hh]: Make radians

File: embedded/servo_bq/set_servo_message.hh [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
+#include <vector>
 
 namespace jet {
 
+// struct ServoCommand{
+// 	int servo_index;
+// 	float target_percentage;
+// 	SERIALIZABLE_STRUCTURE(ServoCommand, servo_index, target_percentage);
+// };
+
```
Comment (jpanikulam) [inline embedded/servo_bq/set_servo_message.hh]: Delete commented

File: embedded/servo_bq/set_servo_message.hh [hunk 5]
```diff
@@ -1,13 +1,23 @@
 #pragma once
 
 #include "infrastructure/comms/schemas/message.hh"
+// #include "infrastructure/comms/serialization/serialization_macros.hh"
```
Comment (jpanikulam) [inline embedded/servo_bq/set_servo_message.hh]: Delete commented

File: embedded/servo_driver/cfg/servo_cfg3.yaml [hunk 6]
```diff
@@ -1,3 +1,4 @@
+index: 3
 calibrated_center: 50
 calibrated_max: 71
 max_angle: 20
```
Comment (jpanikulam) [inline embedded/servo_driver/cfg/servo_cfg3.yaml]: Units if practical

File: embedded/servo_bq/servo_balsaq.cc [hunk 7]
```diff
@@ -1,8 +1,10 @@
 //%bin(servo_balsaq_main)
 #include "embedded/servo_bq/servo_balsaq.hh"
 #include "infrastructure/balsa_queue/bq_main_macro.hh"
+// #include "servo_driver.hh"
```
[the remark this answers is not in the document]
Comment (jpanikulam) [reply embedded/servo_bq/servo_balsaq.cc]: delete comment

File: embedded/servo_driver/servo_driver.hh [hunk 8]
```diff
@@ -15,16 +15,15 @@
...
+  ServoDriver(const std::string &config_path);
   void set_percentage(int percentage);
   int get_percentage() const;
+  int get_servo_index() const;
   void set_angle(float angle);
 
  private:
-  int channel_;
-  const std::shared_ptr<PwmDriver> &pwm_driver_;
+  int servo_index_;
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.hh]: Initialize this to `-1` and assert that it's positive when using it

File: embedded/servo_testing/manual_servo_input_bq.cc [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  set_servo_message.target_radians = target_radians;
+
+  std::vector<int> servo_indices;
+  servo_indices.push_back(servo_index);
+  set_servo_message.servo_indices = servo_indices;
+
+  std::cout << "enter to send command" << std::endl;
+  std::cin.ignore();
+  publisher_->publish(set_servo_message);
+  exit(0); // What's the right way to shutdown a bq?
```
Comment (jpanikulam) [inline embedded/servo_testing/manual_servo_input_bq.cc]: Good question

File: embedded/servo_bq/set_servo_message.hh [hunk 10]
```diff
@@ -1,13 +1,17 @@
...
+#include <vector>
 
 namespace jet {
 
 struct SetServoMessage : Message {
-  float target_angle; //TODO units
+  std::vector<int> servo_indices;
+  std::vector<float> target_radians;
```
Comment (jpanikulam) [inline embedded/servo_bq/set_servo_message.hh]: Suggest: 
```
struct ServoCommand {
    int id;
    double target_rads;
    SERIALIZABLE_STRUCTURE(ServoCommand, id, target_rads)
};
```

```
struct SetServoMessage : Message {
    std::vector<ServoCommand> servo_commands;
    MESSAGE(SetServoMessage, servo_commands);
};
```

That way, the indices and commands are guaranteed to be correspondent

Comment (IJDykeman) [reply embedded/servo_bq/set_servo_message.hh]: Ben and I tried for a while earlier today to get that to work but there is a vector<struct> serialization issue we couldn't figure out in time.

File: embedded/servo_bq/servo_balsaq.cc [hunk 11]
```diff
@@ -14,36 +15,33 @@
...
+    for (int i = 0; i < message.servo_indices.size(); i++) {
+      auto servo_index = message.servo_indices.at(i);
+      auto target_radian = message.target_radians.at(i);
+      assert(servo_index < servos.size());
+      servos.at(servo_index).set_angle_radians(target_radian);
+    }
   }
 }
+// std::cout << "about to command servo" << std::endl;
+// std::cout << " commanded" << std::endl;
```
Comment (jpanikulam) [inline embedded/servo_bq/servo_balsaq.cc]: Delete comments

File: embedded/servo_bq/set_servo_message.hh (hunk 5 above)
Comment (jpanikulam) [reply embedded/servo_bq/set_servo_message.hh]: Delete commented

File: embedded/servo_bq/set_servo_message.hh (hunk 10 above)
Comment (jpanikulam) [reply embedded/servo_bq/set_servo_message.hh]: Alright

File: embedded/servo_driver/servo_driver.cc [hunk 12]
```diff
@@ -1,11 +1,11 @@
 //%deps(yaml-cpp)
 
 #include "embedded/servo_driver/servo_driver.hh"  // driver library
+#include <math.h>                                 // M_PI
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.cc]: (prefer `<cmath>`, since `<math.h>` is a c header

File: embedded/servo_driver/servo_driver.cc [hunk 13]
```diff
@@ -14,42 +14,63 @@ namespace {
...
+  return radians * (180.0 / M_PI);
+}
 }  // namespace
 
-ServoDriver::ServoDriver(const int channel,
-                         const std::shared_ptr<PwmDriver> &pwm_driver,
-                         const std::string &config_path)
-    : channel_(channel), pwm_driver_(pwm_driver), config_path_(config_path) {
+ServoDriver::ServoDriver(const std::string &config_path) : config_path_(config_path) {
+  int i2cHandle = i2c_open("/dev/i2c-1");  // TODO make configurable
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.cc]: Did you figure out if the environment variable should be used?

File: embedded/servo_bq/servo_balsaq.cc [hunk 14]
```diff
@@ -14,36 +15,31 @@
...
-  servo_index = atoi(argv[1]);  // get the servo number in 0...3 from the command line
-
-  int i2cHandle = i2c_open("/dev/i2c-1");
-  if (i2cHandle == -1) {
-    std::cout << "Failed to open i2c" << std::endl;
+void ServoBq::init(int argc, char* argv[]) {
+  assert(argc > 1);
+  subscriber = make_subscriber("servo_command_channel");
+  const int n_servos = argc - 1;
+  // ServoDriver servos [n_servos];
```
Comment (jpanikulam) [inline embedded/servo_bq/servo_balsaq.cc]: Delete commented code

**Pull request merged**